Improve casting of integer routing parameters

markstory opened the pull request:
Allow negative numbers to be cast to integers for typed routing parameters.

Fixes #16640

markstory committed:
Improve casting of integer routing parameters

Allow negative numbers to be cast to integers for typed routing
parameters.

ADmad commented on src/Controller/ControllerFactory.php:
```diff
@@ -268,7 +268,7 @@ protected function coerceStringToType(string $argument, ReflectionNamedType $typ
                 return is_numeric($argument) ? (float)$argument : null;
             case 'int':
-                return ctype_digit($argument) ? (int)$argument : null;
+                return ctype_digit($argument) || filter_var($argument, FILTER_VALIDATE_INT) ? (int)$argument : null;
```
The filter_var() check doesn't cover the ctype_digit() check too?

othercorey replied:
filter_var doesn't validate leading 0's unless you enable octal, but that will re-interpret the value.

ADmad replied:
Unless we have existing tests for casting strings with leading zeros to int I don't think we need to support it.

othercorey replied:
We do, The first test checks `02` which is why filter_var() fails on its own.

ADmad replied:
😢

markstory replied:
Yeah, I'm not sure we want to try and support octals long term. But its there right now :shrug:

othercorey replied:
We don't support octals. It's just casting the string to int which strips the leading 0s.

markstory replied:
The leading 0 handling could be interpreted as octal support though. I'm ok breaking this, it feels like a scenario that isn't going to be common in userland code.

othercorey replied:
I'm ok with it as well. The unit tests were added to support the code we used to validate.

ADmad replied:
Let's remove the `ctype_digit()` check and the test with leading zeros then.

othercorey commented on tests/TestCase/Controller/ControllerFactoryTest.php:
```diff
@@ -757,9 +755,22 @@ public function testInvokePassedParametersCoercion(): void
+                'plugin' => null,
+                'controller' => 'Dependencies',
+                'action' => 'requiredTyped',
+                'pass' => ['1.0', '-1', '0', ''],
```
```suggestion
                'pass' => ['1.0', '-01', '0', ''],
```

ADmad replied:
`-01` shouldn't be coerced to int IMO.

othercorey replied:
Do we break it now?

ADmad replied:
Personally I would prefer a strict behavior but breaking a tested behavior isn't nice in general.

othercorey replied:
Think the reason we supported leading 0's was partly just because we used ctype_digit() and partly because I've seen leading 0's for IDs in URLs and probably thought that was valid. Some places generate large IDs instead of starting off public user IDs at 1.

I don't know if I didn't support negative numbers on purpose, or just happened that way.

markstory committed:
Remove support for integers with leading 0s

The additional ctype_digit logic is more complex than using filter_var()
and I don't think there are many useful scenarios for leading 0s as they
could be ambiguous with octal representations.

See discussion in #16643
